hl7apy/mllp.py

```diff
@@ -71,36 +71,39 @@ def setup(self):
 
     def handle(self):
         end_seq = self.eb + self.cr
-        try:
-            line = self.request.recv(3)
-        except socket.timeout:
-            self.request.close()
-            return
-
-        if line[:1] != self.sb:  # First MLLP char
-            self.request.close()
-            return
 
-        while line[-2:] != end_seq:
+        # 1) each loop through is a single HL7 message with MLLP markers
+        # 2) we leave the loop via a handled or unhandled exception
+        while True:
             try:
-                char = self.rfile.read(1)
-                if not char:
-                    break
-                line += char
+                line = self.rfile.read(3)
             except socket.timeout:
                 self.request.close()
                 return
 
-        message = self._extract_hl7_message(line.decode(self.encoding))
-        if message is not None:
-            try:
-                response = self._route_message(message)
-            except Exception:
+            if line[:1] != self.sb:  # First MLLP char
                 self.request.close()
-            else:
-                # encode the response
-                self.wfile.write(response.encode(self.encoding))
-        self.request.close()
+                return
+
+            while line[-2:] != end_seq:
+                try:
+                    char = self.rfile.read(1)
+                    if not char:
+                        break
```

Comment:
As no more characters can be read from the socket, and we know that the end sequence was not found, shouldn't this just return (or raise an exception) instead of continuing on to extract an hl7 message?

Reply:
Interesting observation. This method has numerous issues, but we just added the loop. If you view the code with `Hide whitespace changes` selected under `Diff settings` you will see how small the actual change is.

Reply:
Another issue with this code is it doesn't handle the case where the message is split across two TCP packets and only some of the data is available at the server.

Reply:
I see, a very small change.